secret_store/src/listener/service_contract_listener.rs

```diff
@@ -0,0 +1,699 @@
+						false => ServiceTask::RestoreServerKey(server_key_id),
+					})
+				},
+				_ => Err("not supported".into()),
+			};
+
+			// process request result
+			match request_result {
+				Ok(_) => processed_requests += 1,
+				Err(_) => {
+					failed_requests += 1;
+					if failed_requests > MAX_FAILED_RETRY_REQUESTS {
+						return Err("too many failed requests".into());
+					}
+				},
+			}
+		}
+
+		Ok(processed_requests)
+	}
+
+	/// Generate server key.
+	fn generate_server_key(data: &Arc<ServiceContractListenerData>, server_key_id: &ServerKeyId, threshold: &H256) -> Result<Public, String> {
+		let threshold_num = threshold.low_u64();
+		if threshold != &threshold_num.into() || threshold_num >= ::std::usize::MAX as u64 {
+			return Err(format!("invalid threshold {:?}", threshold));
+		}
+
+		// key server expects signed server_key_id in server_key_generation procedure
+		// only signer could store document key for this server key later
+		// => this API (server key generation) is not suitable for usage in encryption via contract endpoint
+		let author_key = Random.generate().map_err(|e| format!("{}", e))?;
+		let server_key_id_signature = sign(author_key.secret(), server_key_id).map_err(|e| format!("{}", e))?;
+		data.key_server.generate_key(server_key_id, &server_key_id_signature, threshold_num as usize)
+			.map_err(Into::into)
+	}
+
+	/// Restore server key.
+	fn restore_server_key(data: &Arc<ServiceContractListenerData>, server_key_id: &ServerKeyId) -> Result<Public, String> {
+		data.key_storage.get(server_key_id)
+			.map_err(|e| format!("{}", e))
+			.and_then(|ks| ks.ok_or("missing key".to_owned()))
+			.map(|ks| ks.public)
+	}
+
+	/// Publish server key.
+	fn publish_server_key(data: &Arc<ServiceContractListenerData>, server_key_id: &ServerKeyId, server_key: &Public) -> Result<(), String> {
+		data.contract.publish_server_key(server_key_id, server_key)
+	}
+}
+
+impl Drop for ServiceContractListener {
+	fn drop(&mut self) {
+		if let Some(service_handle) = self.service_handle.take() {
+			self.data.tasks_queue.shutdown();
+			// ignore error as we are already closing
+			let _ = service_handle.join();
+		}
+	}
+}
+
+impl ChainNotify for ServiceContractListener {
+	fn new_blocks(&self, _imported: Vec<H256>, _invalid: Vec<H256>, enacted: Vec<H256>, _retracted: Vec<H256>, _sealed: Vec<H256>, _proposed: Vec<Bytes>, _duration: u64) {
+		let enacted_len = enacted.len();
+		if enacted_len == 0 {
+			return;
+		}
+
+		self.data.contract.update();
+		if !self.data.contract.is_actual() {
+			return;
+		}
+
+		let reason = "enacted.len() != 0; qed";
+		self.process_service_contract_events(
+			enacted.first().expect(reason).clone(),
+			enacted.last().expect(reason).clone());
+
+		// schedule retry if received enough blocks since last retry
+		// it maybe inaccurate when switching syncing/synced states, but that's ok
+		if self.data.last_retry.fetch_add(enacted_len, Ordering::Relaxed) >= RETRY_INTEVAL_BLOCKS {
```

Comment:
some blocks might be retracted at the same time, so the progression might be a bit smaller than assumed here. That said it probably doesn't matter that much here.

Reply:
Right - there's no need to calc exact number of rectracted/enacted blocks here - just a rough estimate